src/Microsoft.SqlTools.ServiceLayer/Connection/ListDatabaseRequestHandler.cs

```diff
@@ -57,43 +61,66 @@ abstract class ListDatabaseRequestHandler<T> : IListDatabaseRequestHandler
     {
         private static readonly string[] SystemDatabases = new string[] { "master", "model", "msdb", "tempdb" };
 
-        public abstract string QueryText { get; }
-
         public ListDatabasesResponse HandleRequest(ISqlConnectionFactory connectionFactory, ConnectionInfo connectionInfo)
         {
             ConnectionDetails connectionDetails = connectionInfo.ConnectionDetails.Clone();
+            var serverConnection = ConnectionService.OpenServerConnection(connectionInfo);
+
+            SMO.Server SQLServer = new SMO.Server(serverConnection);
+
 
-            // Connect to master
-            connectionDetails.DatabaseName = "master";
-            using (var connection = connectionFactory.CreateSqlConnection(ConnectionService.BuildConnectionString(connectionDetails), connectionDetails.AzureAccountToken))
+            List<T> result = new List<T>();
+            ListDatabasesResponse response = new ListDatabasesResponse();
+
+
+            foreach (SMO.Database db in SQLServer.Databases)
             {
-                connection.Open();
-                ListDatabasesResponse response = new ListDatabasesResponse();
-                using (DbCommand command = connection.CreateCommand())
+                if (db.Name != serverConnection.CurrentDatabase)
                 {
-                    command.CommandText = this.QueryText;
-                    command.CommandTimeout = 15;
-                    command.CommandType = CommandType.Text;
-                    using (var reader = command.ExecuteReader())
+                    if(serverConnection.IsContainedAuthentication){ //Check if the user is a contained user. If yes, then don't add the database.
+                        continue;
+                    }
+
+                    ServerConnection dbServerConnection;
+                    dbServerConnection = serverConnection.GetDatabaseConnection(db.Name);
```

Comment:
You can just run the query on the database directly and it'll handle switching as needed